src/metabase/query_processor/util/add_alias_info.clj

```diff
@@ -39,552 +39,413 @@
+                          {:path stage-path, :join-alias join-alias}))))))
+
+(defn- source-table [query stage-path col]
+  (case (:lib/source col)
+    :source/table-defaults        (:table-id col)
+    (:source/joins
+     :source/implicitly-joinable) (or (escaped-join-alias query stage-path (:metabase.lib.join/join-alias col))
+                                      (throw (ex-info "Resolved metadata is missing ::escaped-join-alias" {:col col})))
+    (:source/previous-stage
+     :source/card)                ::source
+    (:source/expressions
+     :source/aggregations
+     :source/native)              ::none))
+
+(defn- update-field-ref [query path field-ref col]
+  (lib/update-options field-ref assoc
+                      ::source-table (source-table query path col)
+                      ::source-alias (escaped-source-alias query path (:lib/source-column-alias col))))
+
+(defn- add-source-aliases [query path stage]
+  (lib.util.match/replace stage
+    ;; don't recurse into the metadata or joins -- [[lib.walk]] will take care of that recursion for us.
+    (_ :guard (constantly (some (set &parents) [:lib/stage-metadata :joins])))
+    &match
+
+    :field
+    (let [col (resolve-field-ref query path &match)]
+      (-> (update-field-ref query path &match col)
+          (lib/update-options assoc ::resolved col)))
+
+    [:expression _opts expression-name]
+    (lib/update-options &match assoc ::source-table ::none)
+
+    :aggregation
+    (lib/update-options &match assoc ::source-table ::none)))
+
+(mu/defn- add-desired-aliases-to-aggregations :- ::lib.schema/stage.mbql
+  "Update any aggregations FIRST so we can use their aliases when updating aggregation refs"
+  [query            :- ::lib.schema/query
+   path             :- ::lib.walk/path
+   returned-columns :- :metabase.lib.metadata.calculation/returned-columns
+   stage            :- ::lib.schema/stage.mbql]
+  (letfn [(resolve-aggregation [[_tag opts, :as ag-clause]]
+            (or (m/find-first #(and (= (:lib/source %) :source/aggregations)
+                                    (= (:lib/source-uuid %) (:lib/uuid opts)))
+                              returned-columns)
+                (throw (ex-info "aggregation definition was not resolved"
+                                {:aggregation ag-clause}))))
+          (update-aggregation [ag-clause]
+            (let [col           (resolve-aggregation ag-clause)
+                  desired-alias (escaped-desired-alias query path (or (:lib/desired-column-alias col)
+                                                                      (throw (ex-info "Resolved aggregation is missing :lib/desired-column-alias"
+                                                                                      {:ag-clause ag-clause, :col col}))))]
+              (lib/update-options ag-clause assoc
+                                  ;; not really convinced it makes sense for an aggregation definition to have a
```

Comment:
Make this a TODO. We'll want to experimentally drop this someday, rather than maintain possibly-dead logic forever.